File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n = len(maps)
    m = len(maps[0])
    answer = bfs(0, 0, n, m, maps)
    return answer

def bfs(r, c, n, m, maps):
    # (r, c, 1)는 현재 위치와 이동한 거리
    q = deque([(r, c, 1)])
    # 이동 방향 정의
    directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
    while q:
        a, b, dist = q.popleft()
        # 도착한 경우
        if a == n-1 and b == m-1:
            return dist
        # 지도의 끝인 경우
        if a < 0 or b < 0 \
        or a >= n or b >= m:
            continue
        # 벽인 경우
        if maps[a][b] == 0:
            continue
        # 방문 처리(벽으로 막기)
        maps[a][b] = 0
        # 상하좌우로 이동
        for x, y in directions:
            q.append((a + x, b + y, dist + 1))
    return -1
            

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    private static int[][] deltas = {{1,0}, {0,1}, {-1,0}, {0,-1}};
    
    private static int n; // 행 길이
    private static int m; // 열 길이
    
    private static int solve(int[][] maps) {
        ArrayDeque<int[]> dq = new ArrayDeque<>();
        boolean[][] visited = new boolean[n][m];
        dq.offer(new int[] {0, 0, 1});
        
        while (!dq.isEmpty()) {
            int[] cur = dq.pollFirst();
            int r = cur[0];
            int c = cur[1];
            int d = cur[2];
            
            // 종료 조건
            if (r == n - 1 && c == m - 1) {
                return d;
            }
            
            // 범위를 벗어나면
            if (r < 0 || c < 0 || r >= n || c >= m) {
                continue;
            }
            
            // 벽이 있으면
            if (maps[r][c] == 0) {
                continue;
            }
            
            // 이미 방문했으면
            if (visited[r][c]) {
                continue;
            }
            
            visited[r][c] = true;
            
            for (int[] delta : deltas) {
                int nr = r + delta[0];
                int nc = c + delta[1];
                dq.addLast(new int[] {nr, nc, d + 1});
            }
        }
    
        return -1;
    }
    public int solution(int[][] maps) {
        n = maps.length;
        m = maps[0].length;
        
        int answer = solve(maps);
        return answer;
    }
}